### ISSUE TITLE
wish: to have pyuv ported to Pypy...

### ISSUE THREAD

**Issue by inercia** [saghul/pyuv#49]
Hi there,

This is just a wish: it would be great to have pyuv working on Pypy. There is no problem compiling and installing it with Pypy, but some unit tests fail:

```
test_fsevent_dir (__main__.FSEventTest) ... ok
test_fsevent_nrefile (__main__.FSEventTest) ... ok
test_fsevent_basic (__main__.FSEventTestBasic) ... ok
test_fspoll1 (__main__.FSPollTest) ... RPython traceback:
  File "pypy_interpreter_gateway.c", line 842, in BuiltinCodePassThroughArguments1_funcrun_obj
  File "pypy_module_cpyext_api_4.c", line 40772, in generic_cpy_call__StdObjSpaceConst_funcPtr_SomeI
  File "pypy_module_cpyext_pyobject.c", line 568, in make_ref
  File "pypy_module_cpyext_pyobject.c", line 1486, in create_ref
Fatal RPython error: AssertionError
Aborted (core dumped)

test_client1 (__main__.TCPErrorTest) ... ok
test_client2 (__main__.TCPErrorTest) ... ok
test_open (__main__.TCPErrorTest) ... ok
test_tcp_flags (__main__.TCPFlagsTest) ... ok
test_tcp_shutdown (__main__.TCPShutdownTest) ... ok
test_tcp1 (__main__.TCPTest) ... ok
test_tcp_write_cancel (__main__.TCPTest2) ... ok
test_tcp_list (__main__.TCPTestList) ... ok
test_tcp_list_null (__main__.TCPTestListNull) ... ok
test_tcp_list_unicode (__main__.TCPTestListUnicode) ... RPython traceback:
  File "pypy_interpreter_gateway.c", line 842, in BuiltinCodePassThroughArguments1_funcrun_obj
  File "pypy_module_cpyext_api_4.c", line 40772, in generic_cpy_call__StdObjSpaceConst_funcPtr_SomeI
  File "pypy_module_cpyext_pyobject.c", line 568, in make_ref
  File "pypy_module_cpyext_pyobject.c", line 1486, in create_ref
Fatal RPython error: AssertionError
Aborted (core dumped)

test_udp_pingpong (__main__.UDPTest) ... RPython traceback:
  File "pypy_interpreter_gateway.c", line 842, in BuiltinCodePassThroughArguments1_funcrun_obj
  File "pypy_module_cpyext_api_4.c", line 40772, in generic_cpy_call__StdObjSpaceConst_funcPtr_SomeI
  File "pypy_module_cpyext_pyobject.c", line 568, in make_ref
  File "pypy_module_cpyext_pyobject.c", line 1486, in create_ref
Fatal RPython error: AssertionError
Aborted (core dumped)
```

Thanks in advance

Alvaro

**Comment by digi604** [comment]
+1

**Comment by veegee** [comment]
This is actually very close to being done, but I'm having trouble finding the time to do anything lately. That'll change in ~1 month, but in the meantime, it would be great if I could get some help. The existing pyuv_cffi code is here: https://github.com/veegee/guv/tree/develop/pyuv_cffi and it contains everything needed for guv to work at the moment.

**Comment by koehlma** [comment]
I have started working on a complete libuv CFFI based wrapper because I really need libuv to work together with pypy. It is not fully pyuv compatible, although it should be possible to write a thin pyuv compatibility layer: https://github.com/koehlma/uv

**Comment by veegee** [comment]
@koehlma, you should be able to copy/paste my incomplete implementation. I didn't bother to finish implementing all features because they were not needed for my event loop, but the project layout certainly allows them to be very easily added in about an afternoon's worth of work.

**Comment by koehlma** [comment]
@veegee, I have already implemented all stuff you had. But I also want to support all the DNS, FS and miscellaneous parts of libuv which requires a little bit more work.

**Comment by saghul** [comment]
I'm sad you chose the LGPL license, that means I cannot look at that code
or integrate it.

**Comment by koehlma** [comment]
Yes, but MIT is no option for me, sorry. Nevertheless it could still be integrated. In that case people who want to use PyPy together with libuv have to accept the conditions of LGPL or otherwise simply use CPython. Which is IMHO better than having no choice at all. Would that be acceptable? However I do not want to start a discussion about MIT vs. LGPL.

**Comment by saghul** [comment]
I do not intend to start such a discussion either, sorry if my comment led you to believe otherwise!

My end goal here is to eventually have both CFFI and C backends (I already refactored the package structure to make room for it, but never found the time and patience to start). Since pyuv already has documentation and tests IMHO it's a good place to make it happen, since those don't have to be rewritten. A new decorator for skipping unimplemented parts of the CFFI backend could be used.

If anyone wants to commit to this efforts, I'm willing to at least helt review the code and make such person a comitter here so that they can maintain that part.

IANAL, but adding LGPL code to the project is not an option since the license that governs the project is MIT and I don't plan on dual-licensing it.

At any rate, thanks a lot for chiming in, and (as one of the libuv maintainers) I wish you good luck in your project!

**Comment by tilgovi** [comment]
I've got a branch that passes the tests on the brand new PyPy3 5.7 release. I'm not 100% sure how these changes feel yet, but should I open a PR for feedback?

**Comment by tilgovi** [comment]
Here is the comparison view:
https://github.com/saghul/pyuv/compare/v1.x...tilgovi:pypy

**Comment by saghul** [comment]
@tilgovi Changes look good! Mind creating a PR? You probably also want to modify the Travis file so tests run on PyPy too :-)

**Comment by tilgovi** [comment]
Travis will have to wait until Travis updates their PyPy version (issue: travis-ci/travis-ci#7507).

I can add it and mark it as allowed to fail, though.

In the meantime, I would love to understand why the `tp_clear` member is `NULL` during `dealloc` for Loops and Handles in the basetype tests. If it were just Handles, I would suspect something with resurrection, but it happens on Loops, too. If there's a bug in cpyext, I would like to report it.

**Comment by tilgovi** [comment]
I can also have Travis for PyPy3 download a portable pypy3 build, but that'll add a lot of noise to the Travis file.

**Comment by saghul** [comment]
> Travis will have to wait until Travis updates their PyPy version (issue: travis-ci/travis-ci#7507).
> 
> I can add it and mark it as allowed to fail, though.
>
> I can also have Travis for PyPy3 download a portable pypy3 build, but that'll add a lot of noise to the Travis file.

That's fine, I don't want to merge a patch which could break over time if we can't test it. IIRC there is ppa with latest PyPy, you could also use that.

> In the meantime, I would love to understand why the tp_clear member is NULL during dealloc for Loops and Handles in the basetype tests. If it were just Handles, I would suspect something with resurrection, but it happens on Loops, too. If there's a bug in cpyext, I would like to report it.

TBH I have no idea what that could be :-S

**Comment by tilgovi** [comment]
I have opened #232 and after more investigation into the issues I feel pretty secure about it.

I think I have figured out why PyPy and CPython differ in the behavior of `tp_clear` with inheritance, but all the standard library modules in CPython invoke the `tp_clear` function by name, rather than by dereference from the type object. So, I feel good about the change here.

Let me know on the PR if there's anything else I can do here or if any of it makes anyone uneasy.
